runtime/compiler/control/CompilationThread.cpp

```diff
@@ -9268,24 +9310,32 @@ TR::CompilationInfo::compilationEnd(J9VMThread * vmThread, TR::IlGeneratorMethod
 
-                        // Storing the uncompressed metadata in both buffer. 
-                        memcpy(uncompressedData, dataStart, dataSize);
-                        memcpy((U_8*)(uncompressedData)+dataSize, codeStart, codeSize);
-                        int uncompressedDataSize = sizeof(J9JITDataCacheHeader)+sizeof(TR_AOTMethodHeader);
-                        int compressedDataSize = deflateBuffer((U_8*)(uncompressedData)+uncompressedDataSize, dataSize+codeSize-uncompressedDataSize, (U_8*)(compressedData)+uncompressedDataSize, Z_DEFAULT_COMPRESSION);
+                        // Copy the metadata and codedata combined into single buffer so that we can compress it together
+                        memcpy(originalData, dataStart, dataSize);
+                        memcpy((U_8*)(originalData)+dataSize, codeStart, codeSize);
+                        // We do not compress the Method Header so that while inflating the data we can utilize the size of the metadata and codedata
+                        // to allocate appropriate size buffer.
+                        int aotMethodHeaderSize = sizeof(J9JITDataCacheHeader)+sizeof(TR_AOTMethodHeader);
+                        int compressedDataSize = deflateBuffer((U_8*)(originalData)+aotMethodHeaderSize, dataSize+codeSize-aotMethodHeaderSize, (U_8*)(compressedData)+aotMethodHeaderSize, Z_DEFAULT_COMPRESSION);
                         if (compressedDataSize != COMPRESSION_FAILED)
                            {
-                           // We are going to store both metadata and code in contiguous memory in shared class cache.
-                           // In load run we get the whole buffer from the cache and use the information from header to
-                           // Get the relocation data and compiled code.
-                           // Because of this reason we copy both metadata and compiled code in one buffer and deflate it together.
+                           /**
+                             * We are going to store both metadata and code in contiguous memory in shared class cache.
+                             * In load run we get the whole buffer from the cache and use the information from header to
+                             * Get the relocation data and compiled code.
+                             * Because of this reason we copy both metadata and compiled code in one buffer and deflate it together.
+                             * TODO: We will always query the shared class cache to get full data for stored AOT compiled method 
+                             * that is combined meta data and code data. Even if compression of AOT bytes is disabled, we do not need
+                             * to send two different buffers to store in cache and also as no one queries either code data / metadata for
+                             * method, clean up the share classs cache API.
```

Comment:
We should open up an issue to track this improvement. @dsouzai FYI.

Reply:
Opened up issue https://github.com/eclipse/openj9/issues/6216